ipa-core/src/protocol/basics/validate.rs

```diff
@@ -0,0 +1,284 @@
+    secret_sharing::{replicated::ReplicatedSecretSharing, SharedValue},
+    seq_join::assert_send,
+};
+
+type HashFunction = Sha256;
+type HashSize = <HashFunction as OutputSizeUser>::OutputSize;
+type HashOutputArray = [u8; <HashSize as Unsigned>::USIZE];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct HashValue(GenericArray<u8, HashSize>);
+
+impl Serializable for HashValue {
+    type Size = HashSize;
+    type DeserializationError = Infallible;
+
+    fn serialize(&self, buf: &mut GenericArray<u8, Self::Size>) {
+        buf.copy_from_slice(self.0.as_slice());
+    }
+
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Result<Self, Self::DeserializationError> {
+        Ok(Self(buf.to_owned()))
+    }
+}
+
+impl Message for HashValue {}
+
+impl From<HashFunction> for HashValue {
+    fn from(value: HashFunction) -> Self {
+        // Ugh: The version of sha2 we currently use doesn't use the same GenericArray version as we do.
+        HashValue(GenericArray::from(<HashOutputArray>::from(
+            value.finalize_fixed(),
+        )))
+    }
+}
+
+struct ReplicatedValidatorFinalization {
+    f: Pin<Box<(dyn Future<Output = Result<(), Error>> + Send)>>,
+}
+
+impl ReplicatedValidatorFinalization {
+    fn new<C: Context + 'static>(active: ReplicatedValidatorActive<C>) -> Self {
+        let ReplicatedValidatorActive {
+            ctx,
+            left_hash,
+            right_hash,
+        } = active;
+        let left_hash = HashValue::from(left_hash);
+        let right_hash = HashValue::from(right_hash);
+        let left_peer = ctx.role().peer(Direction::Left);
+        let right_peer = ctx.role().peer(Direction::Right);
+
+        let f = Box::pin(assert_send(async move {
+            try_join(
+                ctx.send_channel(left_peer)
+                    .send(RecordId::FIRST, left_hash.clone()),
+                ctx.send_channel(right_peer)
+                    .send(RecordId::FIRST, right_hash.clone()),
+            )
+            .await?;
+            let (left_recvd, right_recvd) = try_join(
+                ctx.recv_channel(left_peer).receive(RecordId::FIRST),
+                ctx.recv_channel(right_peer).receive(RecordId::FIRST),
+            )
+            .await?;
+            if left_hash == left_recvd && right_hash == right_recvd {
+                Ok(())
+            } else {
+                Err(Error::Internal) // TODO add a code
+            }
+        }));
+        Self { f }
+    }
+
+    fn poll(&mut self, cx: &mut TaskContext<'_>) -> Poll<Result<(), Error>> {
+        self.f.poll_unpin(cx)
+    }
+}
+
+struct ReplicatedValidatorActive<C> {
+    ctx: C,
+    left_hash: Sha256,
+    right_hash: Sha256,
+}
+
+impl<C: Context + 'static> ReplicatedValidatorActive<C> {
+    fn new(ctx: C) -> Self {
+        Self {
+            ctx,
+            left_hash: HashFunction::new(),
+            right_hash: HashFunction::new(),
+        }
+    }
+
+    fn update<S, V>(&mut self, s: &S)
```

Comment:
For debugging, it might be useful if the update API takes both a `Step` and data, and creates a trace of the validator inputs. Then we can diagnose where things went wrong if there is a mismatch. (Obviously, we would want a flag so we only pay the cost of the detailed tracing when needed.)

Reply:
The context carries a step, so we could smuggle that in somewhere.

I'm more concerned that I've been unable to instantiate this code :)